File: src/main/java/com/gdep/CodeManager.java
package com.gdep;

import com.dslplatform.json.CompiledJson;
import com.dslplatform.json.DslJson;
import com.dslplatform.json.runtime.Settings;
import java.io.BufferedInputStream;
import java.io.BufferedOutputStream;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.OutputStream;
import java.nio.file.FileSystems;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.PathMatcher;
import java.nio.file.StandardOpenOption;
import java.security.DigestInputStream;
import java.security.MessageDigest;
import java.time.Duration;
import java.time.Instant;
import java.util.ArrayList;
import java.util.Collections;
import java.util.HashSet;
import java.util.HexFormat;
import java.util.List;
import java.util.Set;
import java.util.stream.Stream;
import org.gradle.internal.impldep.org.jspecify.annotations.Nullable;
import org.gradle.tooling.GradleConnector;
import org.gradle.tooling.ProjectConnection;
import org.gradle.tooling.model.DomainObjectSet;
import org.gradle.tooling.model.UnsupportedMethodException;
import org.gradle.tooling.model.eclipse.EclipseExternalDependency;
import org.gradle.tooling.model.eclipse.EclipseProject;
import org.gradle.tooling.model.eclipse.EclipseSourceDirectory;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class CodeManager {
    private final String cwd;
    private final String cacheDir;

    // private final Gson gson;
    // private final JsonReader jsonReader;
    private final DslJson<Object> dslJson = new DslJson<>(Settings.withRuntime().includeServiceLoader());

    static final Logger logger = LoggerFactory.getLogger(CodeManager.class);

    static final Duration TOOLING_API_CACHE_TTL = Duration.ofHours(3);

    // ============================
    // SourceCode implementations
    // ============================

    @CompiledJson(onUnknown = CompiledJson.Behavior.FAIL)
    public record ExternalSourceCode(
            String sourceDirPath, List<String> sourceFiles, String sourceJarPath, String sourceJarHash)
            implements SourceCode {

        public ExternalSourceCode {
            sourceFiles = Collections.unmodifiableList(sourceFiles);
        }
    }

    private static record ProjectSourceCode(String sourceDirPath, List<String> sourceFiles) implements SourceCode {}

    // ============================

    // ============================
    // Gradle Tooling outputs
    // ============================

    public static record GradleToolingInfo(
            Set<String> externalSourceJars,
            Set<String> projectSourceDirectories,
            @Nullable String jdkPath) {

        public GradleToolingInfo {
            externalSourceJars = Collections.unmodifiableSet(externalSourceJars);
            projectSourceDirectories = Collections.unmodifiableSet(projectSourceDirectories);
        }
    }

    @CompiledJson(onUnknown = CompiledJson.Behavior.FAIL)
    public static record CachedGradleToolingInfo(
            GradleToolingInfo gradleToolingInfo,
            Set<String> fingerPrintFiles,
            String fingerPrintHash,
            Instant expiresAt) {

        public CachedGradleToolingInfo {
            fingerPrintFiles = Collections.unmodifiableSet(fingerPrintFiles);
        }
    }

    public CodeManager(String cwd, String cacheDir) {
        this.cwd = Util.cleanPath(cwd);
        this.cacheDir = Util.cleanPath(cacheDir);

        // this.gson = new GsonBuilder().setPrettyPrinting().create();
    }

    // ============================

    public List<SourceCode> getExternalSourceCodes() throws IOException {
        GradleToolingInfo info = getGradleToolingInfo();

        List<SourceCode> sourceCodes = new ArrayList<SourceCode>();

        // get SourceCode from externSourceJarPaths
        for (final String jarPath : info.externalSourceJars()) {
            sourceCodes.add(getSourceCodeFromJar(jarPath));
        }

        // get SourceCode from jdk (if it exists)
        if (info.jdkPath() != null) {
            Path jdkZipPath = Path.of(info.jdkPath(), "lib", "src.zip");
            try {
                sourceCodes.add(getSourceCodeFromJar(jdkZipPath.toString()));
            } catch (Exception e) {
                logger.error("failed to get jdk sources from {}", jdkZipPath, e);
            }
        }

        // get SourceCode from projectSourceDirs
        for (final String srcDir : info.projectSourceDirectories()) {
            List<String> files = Util.getFilesInDirectory(srcDir);

            sourceCodes.add(new ProjectSourceCode(srcDir, files));
        }

        return sourceCodes;
    }

    private SourceCode getSourceCodeFromJar(String sourceJarPath) throws IOException {
        // first get hash of the jar
        String sourceJarHash = Util.hashFileToString(sourceJarPath);

        // check cache
        Path jsonCachePath = Path.of(getJsonJarCachePath(sourceJarHash));

        ExternalSourceCode sourceCode;

        if (Files.exists(jsonCachePath) && Files.isRegularFile(jsonCachePath)) {
            // if cache exists, we just parse json cache
            // String jsonString = Files.readString(jsonCachePath);
            // sourceCode = gson.fromJson(jsonString, ExternalSourceCode.class);
            byte[] jsonBytes = Files.readAllBytes(jsonCachePath);
            sourceCode = dslJson.deserialize(ExternalSourceCode.class, jsonBytes, jsonBytes.length);
        } else {
            // if it doesn't exist in cache, create a cache
            String sourceDirPath = getUnzippedSourceDirPath(sourceJarPath, sourceJarHash);

            Util.extractZip(sourceJarPath, sourceDirPath);

            List<String> sourceFiles = Util.getFilesInDirectory(sourceDirPath);

            sourceCode = new ExternalSourceCode(sourceDirPath, sourceFiles, sourceJarPath, sourceJarHash);

            // String jsonString = gson.toJson(sourceCode);
            try (var out = new BufferedOutputStream(
                    Files.newOutputStream(jsonCachePath, StandardOpenOption.WRITE, StandardOpenOption.CREATE_NEW))) {
                dslJson.serialize(sourceCode, out);
            }

            // Files.writeString(jsonCachePath, jsonString, StandardOpenOption.WRITE, StandardOpenOption.CREATE_NEW);
        }

        if (sourceCode == null) {
            throw new RuntimeException("failed to deserialize sourceCode");
        }

        return sourceCode;
    }

    private GradleToolingInfo getGradleToolingInfo() throws IOException {
        // first collect fingerprint hash
        Set<String> fingerPrintFiles = getFingerPrintFiles();

        String hash = getGradleFingerPrintHash(fingerPrintFiles);

        // check cache
        Path jsonCachePath = Path.of(getJsonToolingAPICachePath(hash));

        try {
            if (Files.exists(jsonCachePath) && Files.isRegularFile(jsonCachePath)) {
                // String jsonString = Files.readString(jsonCachePath);
                // CachedGradleToolingInfo cachedInfo = gson.fromJson(jsonString, CachedGradleToolingInfo.class);

                byte[] jsonBytes = Files.readAllBytes(jsonCachePath);
                CachedGradleToolingInfo cachedInfo =
                        dslJson.deserialize(CachedGradleToolingInfo.class, jsonBytes, jsonBytes.length);

                if (cachedInfo == null) {
                    throw new RuntimeException("failed to deserialize cachedInfo");
                }

                // only if fingerPrintHash matches and the cache is not expired yet
                if (cachedInfo.fingerPrintHash().equals(hash) && Instant.now().isBefore(cachedInfo.expiresAt())) {
                    return cachedInfo.gradleToolingInfo();
                }
            }
        } catch (Exception e) {
            logger.error("could not load cache from {}", jsonCachePath, e);
        }

        Set<String> externSourceJarPaths = new HashSet<>();

        String jdkPath = null;

        Set<String> projectSourceDirs = new HashSet<>();

        try (ProjectConnection connection = GradleConnector.newConnector()
                .forProjectDirectory(new File(cwd))
                .connect()) {

            EclipseProject project = connection.getModel(EclipseProject.class);

            // get external source jar paths
            Set<EclipseExternalDependency> deps = getProjectDependencies(project);
            for (final EclipseExternalDependency dep : deps) {
                File sourceFile = dep.getSource();

                if (sourceFile != null) {
                    externSourceJarPaths.add(Util.cleanPath(sourceFile.getCanonicalPath()));
                }
            }

            // try to get jdk path
            // this is technically wrong since each project could have
            // different jdk versions
            //
            // but I don't think it'd matter too much
            //
            // but the correct thing would be to collect java home for every projects
            try {
                var javaSettings = project.getJavaSourceSettings();
                if (javaSettings != null) {
                    jdkPath = Util.cleanPath(javaSettings.getJdk().getJavaHome().getPath());
                }
            } catch (UnsupportedMethodException e) {
                logger.error("failed to get jdk path", e);
            }

            // get project source directories
            // TODO: implement include and exclude patterns if you can
            Set<EclipseSourceDirectory> srcDirs = getProjectSourceDirs(project);
            for (final EclipseSourceDirectory dir : srcDirs) {
                projectSourceDirs.add(Util.cleanPath(dir.getPath()));
            }
        }

        GradleToolingInfo info = new GradleToolingInfo(externSourceJarPaths, projectSourceDirs, jdkPath);

        Instant expiresAt = Instant.now().plus(TOOLING_API_CACHE_TTL);

        CachedGradleToolingInfo cachedInfo = new CachedGradleToolingInfo(info, fingerPrintFiles, hash, expiresAt);

        try (var out = new BufferedOutputStream(Files.newOutputStream(
                jsonCachePath,
                StandardOpenOption.WRITE,
                StandardOpenOption.CREATE,
                StandardOpenOption.TRUNCATE_EXISTING))) {
            dslJson.serialize(cachedInfo, out);
        }

        return info;
    }

    // ============================
    // helper fucntions
    // ============================
    private String getUnzippedSourceDirPath(String sourceJarPath, String sourceJarHash) {
        if (sourceJarPath.endsWith("-sources.jar")) {
            sourceJarPath = sourceJarPath.substring(0, sourceJarPath.length() - "-sources.jar".length());
        } else if (sourceJarPath.endsWith(".jar")) {
            sourceJarPath = sourceJarPath.substring(0, sourceJarPath.length() - ".jar".length());
        } else if (sourceJarPath.endsWith(".zip")) {
            sourceJarPath = sourceJarPath.substring(0, sourceJarPath.length() - ".zip".length());
        }

        Path fileNamePath = Path.of(sourceJarPath).getFileName();

        String dirName;

        if (fileNamePath == null) {
            dirName = "zipped-src-" + sourceJarHash;
        } else {
            dirName = fileNamePath.toString() + "-" + sourceJarHash;
        }

        return Util.cleanPath(Path.of(this.cacheDir, dirName)).toString();
    }

    private String getJsonJarCachePath(String sourceJarHash) {
        String fileName = sourceJarHash + ".jar-cache.json";

        return Util.cleanPath(Path.of(this.cacheDir, fileName)).toString();
    }

    private String getJsonToolingAPICachePath(String fingerPrintHash) {
        String fileName = fingerPrintHash + ".api-cache.json";

        return Util.cleanPath(Path.of(this.cacheDir, fileName)).toString();
    }

    private static Set<EclipseExternalDependency> getProjectDependencies(EclipseProject project) {
        Set<EclipseExternalDependency> deps = new HashSet<>();
        getProjectDependenciesImpl(deps, project);

        return deps;
    }

    private static void getProjectDependenciesImpl(Set<EclipseExternalDependency> deps, EclipseProject project) {
        DomainObjectSet<? extends EclipseExternalDependency> projectDeps = project.getClasspath();

        deps.addAll(projectDeps);

        DomainObjectSet<? extends EclipseProject> children = project.getChildren();

        for (final EclipseProject child : children) {
            getProjectDependenciesImpl(deps, child);
        }
    }

    private static Set<EclipseSourceDirectory> getProjectSourceDirs(EclipseProject project) {
        Set<EclipseSourceDirectory> deps = new HashSet<>();
        getProjectSourceDirsImpl(deps, project);

        return deps;
    }

    private static void getProjectSourceDirsImpl(Set<EclipseSourceDirectory> deps, EclipseProject project) {
        DomainObjectSet<? extends EclipseSourceDirectory> srcDirs = project.getSourceDirectories();

        deps.addAll(srcDirs);

        DomainObjectSet<? extends EclipseProject> children = project.getChildren();

        for (final EclipseProject child : children) {
            getProjectSourceDirsImpl(deps, child);
        }
    }

    private String getGradleFingerPrintHash(Set<String> fingerPrintFiles) throws IOException {
        // sort the files
        List<String> sortedFiles = fingerPrintFiles.stream().sorted().toList();

        MessageDigest digest = Util.getSha256Digest();

        for (String file : sortedFiles) {
            try (var dis = new DigestInputStream(new BufferedInputStream(new FileInputStream(file)), digest)) {
                dis.transferTo(OutputStream.nullOutputStream());
            } catch (Exception e) {
                logger.error("failed open file \"{}\" to caculcate finger printing hash", file, e);
            }
        }

        return HexFormat.of().formatHex(digest.digest());
    }

    // list is copy pasted from
    // https://github.com/gradle/actions/blob/main/sources/src/cache-service-basic.ts
    private static Set<String> fingerPrintFilePatterns = Set.of(
            "glob:**/*.gradle*",
            "glob:**/gradle-wrapper.properties",
            "glob:buildSrc/**/Versions.kt",
            "glob:buildSrc/**/Dependencies.kt",
            "glob:gradle/*.versions.toml",
            "glob:**/versions.properties");

    private Set<String> getFingerPrintFiles() throws IOException {
        List<PathMatcher> matchers = fingerPrintFilePatterns.stream()
                .map(x -> FileSystems.getDefault().getPathMatcher(x))
                .toList();

        Set<String> toReturn = new HashSet<>();

        try (Stream<Path> direntStream = Files.walk(Path.of(cwd))) {
            List<Path> dirents = direntStream.filter(Files::isRegularFile).toList();

            for (final Path dirent : dirents) {
                for (PathMatcher matcher : matchers) {
                    if (matcher.matches(dirent)) {
                        toReturn.add(Util.cleanPath(dirent).toString());
                        break;
                    }
                }
            }
        }

        return toReturn;
    }
}

File: go_wrapper/code_manager.go
package main

import (
	"bytes"
	"crypto/sha256"
	"encoding/hex"
	"encoding/json"
	"fmt"
	"github.com/gobwas/glob"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"time"
)

var ToolingApiCacheTTL = time.Hour * 3

type SourceCode interface {
	GetSourceDirPath() string
	GetSourceFiles() []string
}

type ExternalSourceCode struct {
	SourceDirPath string
	SourceFiles   []string

	SourceJarPath string
	SourceJarHash string

	KnownCacheDir string
}

func (es *ExternalSourceCode) GetSourceDirPath() string {
	return es.SourceDirPath
}

func (es *ExternalSourceCode) GetSourceFiles() []string {
	return es.SourceFiles
}

type ProjectSourceCode struct {
	SourceDirPath string
	SourceFiles   []string
}

func (ps *ProjectSourceCode) GetSourceDirPath() string {
	return ps.SourceDirPath
}

func (ps *ProjectSourceCode) GetSourceFiles() []string {
	return ps.SourceFiles
}

type GradleToolingInfo struct {
	ExternalSourceJars       []string `json:"externalSourceJars"`
	ProjectSourceDirectories []string `json:"projectSourceDirectories"`
	JdkPaths                 []string `json:"jdkPaths"`
}

type CachedGradleToolingInfo struct {
	GradleToolingInfo

	FingerPrintFiles []string
	FingerPrintHash  string
	ExpiresAt        time.Time
}

func GetSourceCodes(ignoreCache bool) ([]SourceCode, error) {
	var projectDir string
	var inGradleProject bool
	var err error

	// try to get project directory
	projectDir, inGradleProject, err = WalkUpToGradleProjectDir()
	if err != nil {
		return nil, err
	}

	if inGradleProject { // we are in gradle project
		info, err := GetGradleToolingInfo(projectDir, ignoreCache)
		if err != nil {
			return nil, err
		}

		var sourceCodes []SourceCode

		// get SourceCode from externSourceJarPaths
		for _, jarPath := range info.ExternalSourceJars {
			sc, err := GetSourceCodeFromJar(jarPath)
			if err != nil {
				return nil, err
			}
			sourceCodes = append(sourceCodes, sc)
		}

		// get SourceCode from jdk (if it exists)
		for _, jdkPath := range info.JdkPaths {
			jdkZipPath := filepath.Join(filepath.Clean(jdkPath), "lib", "src.zip")
			jarSourceCode, err := GetSourceCodeFromJar(jdkZipPath)

			if err != nil {
				ErrLogger.Println(err)
			} else {
				sourceCodes = append(sourceCodes, jarSourceCode)
			}
		}

		// get SourceCode from projectSourceDirs
		for _, srcDir := range info.ProjectSourceDirectories {
			files, err := GetFilesInDirectory(srcDir)
			if err != nil {
				ErrLogger.Printf("failed to get files in \"%s\": %v", srcDir, err)
				continue
			}

			sourceCodes = append(sourceCodes, &ProjectSourceCode{
				SourceDirPath: srcDir,
				SourceFiles:   files,
			})
		}

		return sourceCodes, nil
	} else { // we are not in gradle project
		// then just try to get zipped java src from jdk
		jdkZipPath := filepath.Join(ProgramInfo.JavaHome, "lib", "src.zip")
		jarSourceCode, err := GetSourceCodeFromJar(jdkZipPath)

		if err != nil {
			return nil, fmt.Errorf("failed to get jdk source code: %w", err)
		}

		return []SourceCode{jarSourceCode}, nil
	}
}

func GetGradleToolingInfo(projectDir string, ignoreCache bool) (GradleToolingInfo, error) {
	// first collect fingerprint hash files
	fingerPrintFiles, err := GetFingerPrintFiles(projectDir)
	if err != nil {
		return GradleToolingInfo{}, fmt.Errorf("failed to collect finger print files %w", err)
	}

	// create hash based on finger print files
	hash, err := GetGradleFingerPrintHash(fingerPrintFiles)
	if err != nil {
		return GradleToolingInfo{}, fmt.Errorf("failed to hash finger print files %w", err)
	}

	// check cached json file of GradleToolingInfo
	jsonCachePath := GetJsonToolingAPICachePath(hash)

	if !ignoreCache {
		exists, err := RegularFileExists(jsonCachePath)
		if err != nil {
			return GradleToolingInfo{}, err
		}

		if exists { // cached json file exists, try to use it
			cachedInfo, cacheLoadErr := LoadGradleToolingInfoFromJsonCache(jsonCachePath)

			if cacheLoadErr != nil {
				ErrLogger.Printf("failed to load cached Gradle Tooling API info from \"%s\": %v", jsonCachePath, cacheLoadErr)
				ErrLogger.Printf("trying to get from gradle directly")
			} else if cachedInfo.FingerPrintHash == hash && time.Now().Before(cachedInfo.ExpiresAt) {
				return cachedInfo.GradleToolingInfo, nil
			}
		}
	}

	// either cahce expired or something went wrong
	// get directly from java
	info, err := GetGradleToolingInfoFromJava(projectDir)
	if err != nil {
		return GradleToolingInfo{}, fmt.Errorf("failed to get Gradle Tooling API info from java: %w", err)
	}

	// save it to json cache
	expiresAt := time.Now().Add(ToolingApiCacheTTL)

	cachedInfo := CachedGradleToolingInfo{
		GradleToolingInfo: info,

		FingerPrintFiles: fingerPrintFiles,
		FingerPrintHash:  hash,

		ExpiresAt: expiresAt,
	}

	jsonBytes, err := json.MarshalIndent(cachedInfo, "", "  ")
	if err != nil {
		return GradleToolingInfo{}, err
	}

	err = os.WriteFile(jsonCachePath, jsonBytes, 0664)
	if err != nil {
		return GradleToolingInfo{}, err
	}

	return info, nil
}

func GetSourceCodeFromJar(sourceJarPath string) (SourceCode, error) {
	// first get hash of the jar
	sourceJarHash, err := HashFileToString(sourceJarPath)
	if err != nil {
		return nil, err
	}

	// check cache
	jsonCachePath := GetJsonJarCachePath(sourceJarHash)

	var sourceCode ExternalSourceCode

	fileExists, err := RegularFileExists(jsonCachePath)
	if err != nil {
		return nil, err
	}

	if fileExists { // cache file exists, reusing cache
		sourceCode, err = LoadExternalSourceCodeFromCache(jsonCachePath)
		if err != nil {
			ErrLogger.Printf("failed to load cached source code: %s", err)
			ErrLogger.Printf("falling back to source jar at \"%s\"", sourceJarPath)

			if sourceCode, err = LoadExternalSourceCodeFromJar(sourceJarPath, sourceJarHash, jsonCachePath); err != nil {
				return nil, err
			}
		}
	} else {
		if sourceCode, err = LoadExternalSourceCodeFromJar(sourceJarPath, sourceJarHash, jsonCachePath); err != nil {
			return nil, err
		}
	}

	return &sourceCode, nil
}

func LoadExternalSourceCodeFromCache(jsonCachePath string) (ExternalSourceCode, error) {
	var sourceCode ExternalSourceCode

	jsonBytes, err := os.ReadFile(jsonCachePath)
	if err != nil {
		return ExternalSourceCode{}, err
	}

	err = json.Unmarshal(jsonBytes, &sourceCode)
	if err != nil {
		return ExternalSourceCode{}, err
	}

	// cached ExternalSourceCode's KnownCacheDir doesn't match with actual ProgramInfo.CacheDir.
	// meaning cache directory was moved after the cache was created.
	// so we need to 'repair' the entries in cached sourceCode
	// and resave it.
	if sourceCode.KnownCacheDir != ProgramInfo.CacheDir {
		newSourceFiles := make([]string, len(sourceCode.SourceFiles))

		for i, file := range sourceCode.SourceFiles {
			if !strings.HasPrefix(file, sourceCode.KnownCacheDir) {
				return ExternalSourceCode{}, fmt.Errorf(
					"cached file \"%s\" doesn't start with known cache dir \"%s\"",
					file, sourceCode.KnownCacheDir,
				)
			}

			relPath, err := filepath.Rel(sourceCode.KnownCacheDir, file)
			if err != nil {
				return ExternalSourceCode{}, err
			}

			newPath := filepath.Join(ProgramInfo.CacheDir, relPath)

			newSourceFiles[i] = newPath
		}

		// same with dirs
		var newSourceDirPath string
		{
			if !strings.HasPrefix(sourceCode.SourceDirPath, sourceCode.KnownCacheDir) {
				return ExternalSourceCode{}, fmt.Errorf(
					"cached dir \"%s\" doesn't start with known cache dir \"%s\"",
					sourceCode.SourceDirPath, sourceCode.KnownCacheDir,
				)
			}

			relPath, err := filepath.Rel(sourceCode.KnownCacheDir, sourceCode.SourceDirPath)
			if err != nil {
				return ExternalSourceCode{}, err
			}

			newSourceDirPath = filepath.Join(ProgramInfo.CacheDir, relPath)
		}

		// update source code
		sourceCode.SourceFiles = newSourceFiles
		sourceCode.SourceDirPath = newSourceDirPath
		sourceCode.KnownCacheDir = ProgramInfo.CacheDir

		// update cache in the file
		saveErrMsg := fmt.Sprintf("failed to update external source code to cache \"%s\"", jsonCachePath)
		jsonBytes, err := json.MarshalIndent(sourceCode, "", "  ")
		if err != nil {
			return ExternalSourceCode{}, fmt.Errorf("%s: %w", saveErrMsg, err)
		}

		err = os.WriteFile(jsonCachePath, jsonBytes, 0664)
		if err != nil {
			return ExternalSourceCode{}, fmt.Errorf("%s: %w", saveErrMsg, err)
		}
	}

	return sourceCode, nil
}

func LoadExternalSourceCodeFromJar(
	sourceJarPath,
	sourceJarHash,
	jsonCachePath string,
) (ExternalSourceCode, error) {
	sourceDirPath := GetUnzippedSourceDirPath(sourceJarPath, sourceJarHash)

	// if sourceDirPath already exists, we delete it
	{
		someFile, err := FileExists(sourceDirPath)

		if err != nil {
			return ExternalSourceCode{}, err
		}

		if someFile != nil {
			if !someFile.IsDir() {
				return ExternalSourceCode{}, fmt.Errorf("tried to unzip at \"%s\" but encountered a file. considered deleting it but something other than a directory exists there", sourceDirPath)
			}

			InfoLogger.Printf("trying to unzip source code at \"%s\" but a directory is already there. deleting it", sourceDirPath)

			err := os.RemoveAll(sourceDirPath)
			if err != nil {
				return ExternalSourceCode{}, err
			}
		}
	}

	// unzipping takes fucking forever
	// atleast log it
	InfoLogger.Printf("unzipping \"%s\" to \"%s\"", sourceJarPath, sourceDirPath)
	err := ExtractZip(sourceJarPath, sourceDirPath)
	if err != nil {
		return ExternalSourceCode{}, err
	}

	sourceFiles, err := GetFilesInDirectory(sourceDirPath)
	if err != nil {
		return ExternalSourceCode{}, err
	}

	var sourceCode ExternalSourceCode

	sourceCode = ExternalSourceCode{
		SourceDirPath: sourceDirPath,
		SourceFiles:   sourceFiles,
		SourceJarPath: sourceJarPath,
		SourceJarHash: sourceJarHash,

		KnownCacheDir: ProgramInfo.CacheDir,
	}

	jsonBytes, err := json.MarshalIndent(sourceCode, "", "  ")
	if err != nil {
		return ExternalSourceCode{}, err
	}

	err = os.WriteFile(jsonCachePath, jsonBytes, 0664)
	if err != nil {
		return ExternalSourceCode{}, err
	}

	return sourceCode, nil
}

func LoadGradleToolingInfoFromJsonCache(jsonCachePath string) (CachedGradleToolingInfo, error) {
	jsonBytes, err := os.ReadFile(jsonCachePath)
	if err != nil {
		return CachedGradleToolingInfo{}, err
	}

	var cachedInfo CachedGradleToolingInfo
	err = json.Unmarshal(jsonBytes, &cachedInfo)
	if err != nil {
		return CachedGradleToolingInfo{}, err
	}

	return cachedInfo, nil
}

func GetGradleToolingInfoFromJava(projectDir string) (GradleToolingInfo, error) {
	var args []string

	javaExe := filepath.Join(ProgramInfo.JavaHome, "bin", "java")
	gdepJarPath := filepath.Join(ProgramInfo.ExecDir, "gdep.jar")

	// read about --enable-native-access=ALL-UNNAMED
	// here: https://teamdev.com/jxbrowser/blog/native-access-restrictions-in-java-24/
	if ProgramInfo.JavaVersion >= 24 {
		args = append(
			args,
			"--enable-native-access=ALL-UNNAMED",
		)
	}

	args = append(
		args,
		fmt.Sprintf("-Dgdep.internal.project.dir=%s", projectDir),
		"-jar",
		gdepJarPath,
	)

	args = append(
		args,
		os.Args[1:]...,
	)

	buf := &bytes.Buffer{}

	cmd := exec.Command(
		javaExe,
		args...,
	)

	cmd.Stdout = buf
	cmd.Stderr = os.Stderr

	err := cmd.Run()
	if err != nil {
		return GradleToolingInfo{}, fmt.Errorf("failed to get Gradle Tooling API info from java: %w", err)
	}

	var info GradleToolingInfo
	err = json.Unmarshal(buf.Bytes(), &info)
	if err != nil {
		return GradleToolingInfo{}, fmt.Errorf("failed to get Gradle Tooling API info from java: %w", err)
	}

	return info, nil
}

// ============================
// helper fucntions
// ============================

func GetUnzippedSourceDirPath(sourceJarPath, sourceJarHash string) string {
	filename := filepath.Base(sourceJarPath)

	if cut, didCut := strings.CutSuffix(filename, "-sources.jar"); didCut {
		filename = cut
	} else if cut, didCut := strings.CutSuffix(filename, ".jar"); didCut {
		filename = cut
	} else if cut, didCut := strings.CutSuffix(filename, ".zip"); didCut {
		filename = cut
	}

	var dirName string

	if filename == "." || filename == "" {
		dirName = "zipped-src-" + sourceJarHash
	} else {
		dirName = filename + "-" + sourceJarHash
	}

	return filepath.Join(ProgramInfo.CacheDir, dirName)
}

func GetJsonJarCachePath(sourceJarHash string) string {
	fileName := sourceJarHash + ".jar-cache.json"

	return filepath.Join(ProgramInfo.CacheDir, fileName)
}

func GetJsonToolingAPICachePath(fingerPrintHash string) string {
	fileName := fingerPrintHash + ".api-cache.json"

	return filepath.Join(ProgramInfo.CacheDir, fileName)
}

// list is copy pasted from
// https://github.com/gradle/actions/blob/main/sources/src/cache-service-basic.ts
var fingerPrintFilePatterns = []string{
	"**/*.gradle*",
	"*.gradle*",
	"**/gradle-wrapper.properties",
	"gradle-wrapper.properties",
	"buildSrc/**/Versions.kt",
	"buildSrc/**/Dependencies.kt",
	"gradle/*.versions.toml",
	"**/versions.properties",
	"versions.properties",
}

var fingerPrintFilePatternGlobs []glob.Glob = nil

func GetFingerPrintFiles(projectDir string) ([]string, error) {
	if fingerPrintFilePatternGlobs == nil {
		for _, p := range fingerPrintFilePatterns {
			g := glob.MustCompile(p, '/')
			fingerPrintFilePatternGlobs = append(fingerPrintFilePatternGlobs, g)
		}
	}

	files, err := GetFilesInDirectory(projectDir)
	if err != nil {
		return nil, err
	}

	var fingerPrintFiles []string

	for _, file := range files {
		rel, err := filepath.Rel(projectDir, file)
		if err != nil {
			return nil, err
		}

		rel = strings.ReplaceAll(rel, "\\", "/")

		for _, g := range fingerPrintFilePatternGlobs {
			if g.Match(rel) {
				fingerPrintFiles = append(fingerPrintFiles, file)
				break
			}
		}
	}

	return fingerPrintFiles, nil
}

func GetGradleFingerPrintHash(fingerPrintFiles []string) (string, error) {
	hasher := sha256.New()

	for _, filename := range fingerPrintFiles {
		file, err := os.Open(filename)
		if err != nil {
			return "", err
		}
		defer file.Close()

		_, err = io.Copy(hasher, file)
		if err != nil {
			return "", err
		}
	}

	return hex.EncodeToString(hasher.Sum(nil)), nil
}

// if these files exist, it's either a gradle project
// or a gradle sub project
var gradleMarkers = []string{
	"build.gradle",
	"build.gradle.kts",
	"settings.gradle",
	"settings.gradle.kts",
}

// walks up from current directory to gradle project dir
func WalkUpToGradleProjectDir() (string, bool, error) {
	dir := ProgramInfo.Cwd

	for {
		entries, err := os.ReadDir(dir)
		if err != nil {
			return "", false, fmt.Errorf("failed to walk up the directory: %w", err)
		}

		for _, entry := range entries {
			if !entry.Type().IsRegular() {
				continue
			}

			for _, marker := range gradleMarkers {
				if entry.Name() == marker {
					return dir, true, nil
				}
			}
		}

		// walk up
		walkedUp := filepath.Dir(dir)

		if walkedUp == dir ||
			walkedUp == "." ||
			walkedUp == "" {
			break
		}

		dir = walkedUp
	}

	return "", false, nil
}
